File: backend/src/main/java/observer/quantum/worm/user/UserService.java
package observer.quantum.worm.user;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Random;
import java.util.UUID;
import org.springframework.security.authentication.RememberMeAuthenticationToken;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContext;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.oauth2.client.authentication.OAuth2AuthenticationToken;
import org.springframework.security.oauth2.client.registration.ClientRegistration;
import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository;
import org.springframework.security.oauth2.core.user.OAuth2User;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class UserService {

  private final UserRepository userRepository;
  private final UserIdentityRepository userIdentityRepository;
  private final PasswordEncoder passwordEncoder;
  private final ClientRegistrationRepository clientRegistrationRepository;

  public UserService(
      UserRepository userRepository,
      UserIdentityRepository userIdentityRepository,
      PasswordEncoder passwordEncoder,
      ClientRegistrationRepository clientRegistrationRepository) {
    this.userRepository = userRepository;
    this.userIdentityRepository = userIdentityRepository;
    this.passwordEncoder = passwordEncoder;
    this.clientRegistrationRepository = clientRegistrationRepository;
  }

  public Optional<User> getCurrentUser() {
    SecurityContext securityContext = SecurityContextHolder.getContext();
    Authentication authentication = securityContext.getAuthentication();

    if (authentication instanceof OAuth2AuthenticationToken) {
      var clientRegistrationId =
          ((OAuth2AuthenticationToken) authentication).getAuthorizedClientRegistrationId();
      var userIdentity =
          userIdentityRepository.findByProviderAndProviderId(
              clientRegistrationId, authentication.getName());
      return userIdentity
          .map(UserIdentity::getUser)
          .or(
              () ->
                  Optional.ofNullable(
                      registerUser(
                          clientRegistrationId,
                          ((OAuth2AuthenticationToken) authentication).getPrincipal())));
    } else if (authentication instanceof RememberMeAuthenticationToken) {
      return this.getUser(authentication.getName());
    } else if (authentication instanceof UsernamePasswordAuthenticationToken) {
      return this.getUser(authentication.getName());
    }
    return Optional.empty();
  }

  public User registerUser(String provider, OAuth2User user) {
    var userRecord = new User();
    var userIdentity = new UserIdentity();
    userIdentity.setProvider(provider);
    userIdentity.setProviderId(user.getName());
    userIdentity.setUser(userRecord);
    switch (provider) {
      case "github" -> {
        userIdentity.setProviderUser(user.getAttribute("login"));
        userIdentity.setProviderDisplay(user.getAttribute("name"));
        userRecord.setUsername(user.getAttribute("login"));
        userRecord.setName(user.getAttribute("name"));
        userRecord.setAvatarImageUrl(user.getAttribute("avatar_url"));
        userRecord.setEmail(user.getAttribute("email"));
      }
      case "google" -> {
        userRecord.setName(user.getAttribute("name"));
        userRecord.setUsername(user.getAttribute("email"));
      }
      default -> throw new UserAuthInvalidException();
    }
    // Set a random password for OAuth2 users
    String randomPassword = UUID.randomUUID().toString();
    userRecord.setPassword(passwordEncoder.encode(randomPassword));

    userRepository.save(userRecord);
    userIdentityRepository.save(userIdentity);
    return userRecord;
  }

  public User registerUser(String username, String email, String password, String name) {
    if (userRepository.findByUsername(username).isPresent()) {
      throw new IllegalArgumentException("Username already exists");
    }

    if (!isPasswordStrong(password)) {
      throw new IllegalArgumentException("Password does not meet strength requirements");
    }

    var userRecord = new User();
    userRecord.setUsername(username);
    userRecord.setName(name);
    userRecord.setEmail(email);
    userRecord.setPassword(passwordEncoder.encode(password));
    userRecord.setEnabled(true);

    return userRepository.save(userRecord);
  }

  public Optional<User> getUser(String username) {
    return userRepository.findByUsername(username);
  }

  @Transactional
  public User updateUserDetails(UpdateUserRequest updateUserDto) {
    if (updateUserDto.getName() == null && updateUserDto.getAvatarImageUrl() == null) {
      throw new IllegalArgumentException("At least one field must be provided for update");
    }
    User currentUser = getCurrentUser().orElseThrow(UserAuthInvalidException::new);

    if (updateUserDto.getName() != null) {
      currentUser.setName(updateUserDto.getName());
    }
    if (updateUserDto.getAvatarImageUrl() != null) {
      currentUser.setAvatarImageUrl(updateUserDto.getAvatarImageUrl());
    }

    return userRepository.save(currentUser);
  }

  @Transactional
  public void updatePassword(String oldPassword, String newPassword) {
    User currentUser = getCurrentUser().orElseThrow(UserAuthInvalidException::new);

    if (!passwordEncoder.matches(oldPassword, currentUser.getPassword())) {
      throw new IllegalArgumentException("Old password is incorrect");
    }

    if (!isPasswordStrong(newPassword)) {
      throw new IllegalArgumentException("New password does not meet strength requirements");
    }

    currentUser.setPassword(passwordEncoder.encode(newPassword));
    userRepository.save(currentUser);
  }

  @Transactional
  public void deleteAccount() {
    User currentUser = getCurrentUser().orElseThrow(UserAuthInvalidException::new);
    userIdentityRepository.deleteAllByUser(currentUser);
    userRepository.delete(currentUser);
    SecurityContextHolder.clearContext();
  }

  private boolean isPasswordStrong(String password) {
    // Password must be at least 8 characters long and contain at least one uppercase letter,
    // one lowercase letter, one digit, and one special character
    String passwordRegex = "^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[@$!%*?&])[A-Za-z\\d@$!%*?&]{8,}$";
    return password.matches(passwordRegex);
  }

  @Transactional
  public void migrateUnhashedPasswords() {
    userRepository
        .findAll()
        .forEach(
            user -> {
              if (!user.getPassword()
                  .startsWith("$2a$")) { // Check if the password is not already hashed
                user.setPassword(passwordEncoder.encode(user.getPassword()));
                userRepository.save(user);
              }
            });
  }

  public UsernameAvailabilityResponse checkUsernameAvailability(String username) {
    boolean isAvailable = !userRepository.findByUsername(username).isPresent();
    List<String> suggestions = new ArrayList<>();

    if (!isAvailable) {
      suggestions = generateUsernameSuggestions(username);
    }

    return new UsernameAvailabilityResponse(isAvailable, suggestions);
  }

  private List<String> generateUsernameSuggestions(String username) {
    List<String> suggestions = new ArrayList<>();
    Random random = new Random();

    // Add a random number to the end
    suggestions.add(username + random.nextInt(1000));

    // Add an underscore and a random number
    suggestions.add(username + "_" + random.nextInt(100));

    // Add a random adjective before the username
    String[] adjectives = {"cool", "awesome", "super", "mega", "ultra"};
    suggestions.add(adjectives[random.nextInt(adjectives.length)] + "_" + username);

    // Add a random suffix
    String[] suffixes = {"dev", "coder", "ninja", "guru", "pro"};
    suggestions.add(username + "_" + suffixes[random.nextInt(suffixes.length)]);

    // Ensure all suggestions are unique and available
    return suggestions.stream()
        .distinct()
        .filter(suggestion -> !userRepository.findByUsername(suggestion).isPresent())
        .limit(3)
        .toList();
  }

  public List<String> getEnabledOAuth2Providers() {
    List<String> providers = new ArrayList<>();

    String[] providerIds = {"github", "google"};
    for (String providerId : providerIds) {
      ClientRegistration registration =
          clientRegistrationRepository.findByRegistrationId(providerId);
      if (registration != null) {
        providers.add(registration.getRegistrationId());
      }
    }

    return providers;
  }
}

File: worm-server/src/user/user.service.ts
import {
  Injectable,
  BadRequestException,
} from '@nestjs/common';
import { ValidationException } from '../common/exceptions/validation.exception.js';
import { UserRepository } from './user.repository.js';
import { UserDto } from './user.dto.js';
import { UserEntity } from './user.entity.js';

@Injectable()
export class UserService {
  constructor(
    private readonly userRepo: UserRepository,
  ) {}

  private async validateUserInput(
    username: string,
    password: string,
  ): Promise<void> {
    const errors: Record<string, string[]> = {};

    const existing = await this.userRepo.findByUsername(username);
    if (existing) {
      errors['username'] = ['Username already exists'];
    }

    if (!this.isPasswordStrong(password)) {
      errors['password'] = [
        'Password must contain at least 8 characters',
        'Password must contain at least one uppercase letter',
        'Password must contain at least one lowercase letter',
        'Password must contain at least one number',
        'Password must contain at least one special character',
      ];
    }

    if (Object.keys(errors).length > 0) {
      throw new ValidationException(errors);
    }
  }

  async registerUser(
    username: string,
    email: string,
    password: string,
    name?: string,
  ): Promise<UserEntity> {
    await this.validateUserInput(username, password);

    const hashedPassword = await Bun.password.hash(password);

    const user = new UserEntity({
      username: username,
      email: email,
      password: hashedPassword,
      name: name,
      enabled: true,
    });

    return this.userRepo.createUser(user);
  }

  async getCurrentUser(userId: string): Promise<UserEntity | null> {
    return this.userRepo.findById(userId);
  }

  async updateUserDetails(
    userId: string,
    dto: Partial<UserDto>,
  ): Promise<UserEntity> {
    const user = await this.getCurrentUser(userId);
    if (!user) {
      throw new BadRequestException('User not found');
    }

    const updates: Partial<UserEntity> = {};
    if (dto.name !== undefined) updates.name = dto.name;
    if (dto.avatarImageUrl !== undefined) updates.avatarImageUrl = dto.avatarImageUrl;

    return this.userRepo.updateUser(userId, updates);
  }

  async updatePassword(
    userId: string,
    oldPassword: string,
    newPassword: string,
  ): Promise<void> {
    const user = await this.getCurrentUser(userId);
    if (!user) {
      throw new BadRequestException('User not found');
    }

    const passwordMatches = await Bun.password.verify(oldPassword, user.password);
    if (!passwordMatches) {
      throw new BadRequestException('Old password is incorrect');
    }

    if (!this.isPasswordStrong(newPassword)) {
      throw new ValidationException({
        newPassword: [
          'Password must contain at least 8 characters',
          'Password must contain at least one uppercase letter',
          'Password must contain at least one lowercase letter',
          'Password must contain at least one number',
          'Password must contain at least one special character',
        ],
      });
    }

    const hashedPassword = await Bun.password.hash(newPassword);
    await this.userRepo.updateUser(userId, { password: hashedPassword });
  }

  async deleteAccount(userId: string): Promise<void> {
    await this.userRepo.delete(userId);
  }

  async findByGithubId(githubId: string): Promise<UserEntity | null> {
    return this.userRepo.findByGithubId(githubId);
  }

  async createGithubUser(userData: {
    githubId: string;
    username: string;
    email: string;
    name?: string;
    avatarImageUrl?: string;
  }): Promise<UserEntity> {
    // Check if username already exists
    const existingUser = await this.userRepo.findByUsername(userData.username);

    if (existingUser) {
      // If username exists, append a unique identifier
      userData.username = `${userData.username}_${Date.now()}`;
    }

    const user = new UserEntity({
      username: userData.username,
      email: userData.email,
      name: userData.name || null,
      avatarImageUrl: userData.avatarImageUrl || null,
      githubId: userData.githubId,
      enabled: true,
      password: null, // GitHub users don't have a local password
    });

    return this.userRepo.createUser(user);
  }

  private isPasswordStrong(password: string): boolean {
    const passwordRegex =
      /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&]).{8,}$/;
    return passwordRegex.test(password);
  }

  async checkUsernameAvailability(username: string): Promise<{
    available: boolean;
    suggestions: string[];
  }> {
    const available = await this.userRepo.isUsernameAvailable(username);

    if (available) {
      return {
        available: true,
        suggestions: []
      };
    }

    // Generate suggestions by appending numbers
    const suggestions = Array.from({ length: 3 }, (_, i) =>
      `${username}${i + 1}`
    );

    return {
      available: false,
      suggestions
    };
  }
}
